### PR TITLE
store/copr: set upper limit for extra concurrency

### LINKED ISSUE #41134: Too many cop worker goroutines cause TiDB OOM
## Bug Report

Please answer these questions before submitting your issue. Thanks!

I handled a workload, in which cop client spawned too many extra concurrency workers. It might needs an upper of concurrency.

### 4. What is your TiDB version? (Required)

v6.5

### PR DESCRIPTION
**Pull request by you06**
Signed-off-by: you06 <you1474600@gmail.com>

### What problem does this PR solve?

Issue Number: close #41134

Problem Summary:

### What is changed and how it works?

To speed up the [fanout case](https://github.com/pingcap/tidb/issues/37724), extra workers are created, in a large table lookup case, the number of workers is uncontrollable. This PR sets an upper limit for it.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
store/copr: set upper limit for extra concurrency to prevent OOM
```

### CODE REVIEW

File: store/copr/coprocessor.go [hunk 1]
```diff
@@ -72,6 +73,8 @@ const (
 	smallTaskSigma         = 0.5
 )
 
+var smallTaskConcurrencyLimit = 20 * runtime.NumCPU()
```
Comment (hawkingrei) [inline store/copr/coprocessor.go:76]: It is dangerous. If tidb is in container. The runtime.NumCPU() will return the cpu number of the physical machine. In fact, you need to the value in the cgroup.

Comment (hawkingrei) [reply store/copr/coprocessor.go:76]: Please use ```runtime.GOMAXPROCS(0)``` after ```init``` have been called.

File: store/copr/coprocessor.go [hunk 2]
```diff
@@ -70,13 +70,15 @@ const (
 	copNextMaxBackoff      = 20000
 	CopSmallTaskRow        = 32 // 32 is the initial batch size of TiKV
 	smallTaskSigma         = 0.5
+	smallConcPerCore       = 20
```
Comment (cfzjywxk) [inline store/copr/coprocessor.go:73]: Any experience or experiments about this choice?

File: store/copr/store.go [hunk 3]
```diff
@@ -109,6 +110,7 @@ func (s *Store) GetClient() kv.Client {
 	return &CopClient{
 		store:           s,
 		replicaReadSeed: s.nextReplicaReadSeed(),
+		numcpu:          runtime.GOMAXPROCS(0),
```
Comment (sticnarf) [inline store/copr/store.go:113]: Could it cause contention on the scheduler mutex?

File: store/copr/coprocessor.go (hunk 2 above)
Comment (sticnarf) [reply store/copr/coprocessor.go:73]: With the PR itself, this number seems a bit conservative to me. Maybe ok if this is combined with coprocessor batch.

File: store/copr/store.go (hunk 3 above)
Comment (you06) [reply store/copr/store.go:113]: Yeah, moved this call to `Store`

File: store/copr/coprocessor.go (hunk 2 above)
Comment (you06) [reply store/copr/coprocessor.go:73]: As the solution for the massive table lookup tasks, it's conservative, however, if we calculate the number of goroutines(given a 16c instance, there will be 320 goroutines per cop iterator, and there are 5 table workers in 1 index lookup executor), it's not a small number. The rest speedup operation might be done by coprocessor batch.

**Review by sticnarf — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**